flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -1399,6 +1399,33 @@ public void notifyCheckpointAborted(final long checkpointID) {
         }
     }
 
+    public void notifyCheckpointSubsumed(long checkpointId) {
+        AbstractInvokable invokable = this.invokable;
+
+        if (executionState == ExecutionState.RUNNING && invokable != null) {
+
+            try {
+                invokable.notifyCheckpointSubsumedAsync(checkpointId);
+            } catch (RejectedExecutionException ex) {
+                // This may happen if the mailbox is closed. It means that the task is shutting
+                // down, so we just ignore it.
+                LOG.debug(
+                        "Notify checkpoint subsume {} for {} ({}) was rejected by the mailbox",
+                        checkpointId,
+                        taskNameWithSubtask,
+                        executionId);
+            } catch (Exception e) {
+                if (getExecutionState() == ExecutionState.RUNNING) {
+                    LOG.warn("Error while subsuming checkpoint {}.", checkpointId, e);
+                }
+            }
+        } else {
+            LOG.info(
+                    "Ignoring checkpoint subsume notification for non-running task {}.",
+                    taskNameWithSubtask);
+        }
+    }
```

Comment:
You have resolved my previous comment about code duplication, but it still looks very similar as `notifyCheckpointAborted()` and `notifyCheckpointCompleted()`?

Reply:
I mistakenly thought that you left the comments on `SubtaskCheckpointCoordinatorImpl` class, and I refactored that class to remove code duplication. And after your remider, I see your comments are under `Task` class, and I will refactor later.